File: src/main/scala/fmvpu/core/Params.scala
package fmvpu.core

// Import circe libraries for JSON parsing
import io.circe._
import io.circe.parser._
import io.circe.generic.auto._
import scala.io.Source

/** Configuration parameters for the FMVPU system.
  *
  * This case class contains all the parameters needed to configure an FMVPU instance,
  * including grid dimensions, memory sizes, and network configuration. Parameters are
  * typically loaded from JSON configuration files.
  *
  * @param nChannels Number of parallel network channels between lanes
  * @param width Data width in bits for all data paths
  * @param networkMemoryDepth Maximum depth of output buffers in network nodes
  * @param nDRF Number of registers in the distributed register file per lane
  * @param ddmBankDepth Number of entries per bank in the distributed data memory
  * @param ddmNBanks Number of banks in the distributed data memory
  * @param ddmAddrWidth Address width in bits for distributed data memory
  * @param depthNetworkConfig Number of entries in the network configuration table
  * @param nColumns Number of columns in the LaneGrid
  * @param nRows Number of rows in the LaneGrid
  * @param maxPacketLength Maximum packet length for network switching
  */
case class FMVPUParams(
  nChannels: Int,
  width: Int,
  networkMemoryDepth: Int,
  nDRF: Int,
  ddmBankDepth: Int,
  ddmNBanks: Int,
  ddmAddrWidth: Int,
  depthNetworkConfig: Int,
  nColumns: Int,
  nRows: Int,
  maxPacketLength: Int,
  maxNetworkControlDelay: Int,
  nSlowNetworkControlSlots: Int,
  nFastNetworkControlSlots: Int,
  networkIdentWidth: Int,
) {
  // Calculated parameters based on actual control structure sizes
  import chisel3.util.log2Ceil
  
  // Calculate bit widths for control structures
  private def bitsForChannelSlowControl: Int = {
    1 + // IsPacketMode
    4 * log2Ceil(networkMemoryDepth + 1) + // delays
    1 + // isOutputDelay  
    4 * nChannels + // drive enables
    2 * log2Ceil(maxNetworkControlDelay + 1) + // input sel delays
    2 * log2Ceil(maxNetworkControlDelay + 1) // crossbar sel delays
  }
  
  private def bitsForGeneralSlowControl: Int = {
    2 * log2Ceil(maxNetworkControlDelay + 1) // drf and ddm sel delays
  }
  
  private def bitsForChannelFastControl: Int = {
    2 + // nsInputSel + weInputSel
    2 * log2Ceil(nChannels + 2) // crossbar selections
  }
  
  private def bitsForGeneralFastControl: Int = {
    2 * log2Ceil(nChannels * 2) // drf and ddm selections
  }
  
  val wordsPerChannelSlowNetworkControl = (bitsForChannelSlowControl + width - 1) / width
  val wordsPerGeneralSlowNetworkControl = (bitsForGeneralSlowControl + width - 1) / width  
  val wordsPerChannelFastNetworkControl = (bitsForChannelFastControl + width - 1) / width
  val wordsPerGeneralFastNetworkControl = (bitsForGeneralFastControl + width - 1) / width
  
  // Round slot sizes up to powers of 2 for cleaner addressing
  private val rawWordsPerSlowSlot = wordsPerGeneralSlowNetworkControl + nChannels * wordsPerChannelSlowNetworkControl
  val wordsPerSlowNetworkControlSlot = 1 << scala.math.ceil(scala.math.log(rawWordsPerSlowSlot) / scala.math.log(2)).toInt
  
  private val rawWordsPerFastSlot = wordsPerGeneralFastNetworkControl + nChannels * wordsPerChannelFastNetworkControl  
  val wordsPerFastNetworkControlSlot = 1 << scala.math.ceil(scala.math.log(rawWordsPerFastSlot) / scala.math.log(2)).toInt
  
  // Calculate fastNetworkControlOffset as total slow control space
  val fastNetworkControlOffset = nSlowNetworkControlSlots * wordsPerSlowNetworkControlSlot
  
  val networkControlAddrWidth = scala.math.max(1, 
    32 - Integer.numberOfLeadingZeros(fastNetworkControlOffset + nFastNetworkControlSlots * wordsPerFastNetworkControlSlot - 1)
  )
}

/** Companion object for FMVPUParams with factory methods. */
object FMVPUParams {

  /** Load FMVPU parameters from a JSON configuration file.
    *
    * @param fileName Path to the JSON configuration file
    * @return FMVPUParams instance with configuration loaded from file
    * @throws RuntimeException if the file cannot be parsed or contains invalid parameters
    *
    * @example
    * {{{
    * val params = FMVPUParams.fromFile("config/default.json")
    * val grid = Module(new LaneGrid(params))
    * }}}
    */
  def fromFile(fileName: String): FMVPUParams = {
    val jsonContent = Source.fromFile(fileName).mkString;
    val paramsResult = decode[FMVPUParams](jsonContent);
    paramsResult match {
      case Right(params) =>
        return params;
      case Left(error) =>
        println(s"Failed to parse JSON: ${error.getMessage}")
        System.exit(1)
        null
    }
  }

}

File: python/fmvpu/params.py
from dataclasses import dataclass
import json
import math
from typing import Dict, Any
from fmvpu.test_utils import clog2


@dataclass
class FMVPUParams:
    """Python mirror of the Scala FMVPUParams case class with snake_case naming."""
    
    n_channels: int
    width: int
    # Max depth of the output buffers in the network node.
    network_memory_depth: int
    # Number of registers in the distributed register file.
    n_drf: int
    # Number of vectors stored in the distributed data memory.
    ddm_bank_depth: int
    ddm_n_banks: int
    ddm_addr_width: int
    # Number of entries in the network configuration.
    depth_network_config: int
    n_columns: int
    n_rows: int
    max_packet_length: int
    max_network_control_delay: int
    n_slow_network_control_slots: int
    n_fast_network_control_slots: int
    network_ident_width: int

    # Map camelCase JSON field names to snake_case Python field names
    _FIELD_MAPPING = {
        'nChannels': 'n_channels',
        'width': 'width',
        'networkMemoryDepth': 'network_memory_depth',
        'nDRF': 'n_drf',
        'ddmBankDepth': 'ddm_bank_depth',
        'ddmNBanks': 'ddm_n_banks',
        'ddmAddrWidth': 'ddm_addr_width',
        'depthNetworkConfig': 'depth_network_config',
        'nColumns': 'n_columns',
        'nRows': 'n_rows',
        'maxPacketLength': 'max_packet_length',
        'maxNetworkControlDelay': 'max_network_control_delay',
        'nSlowNetworkControlSlots': 'n_slow_network_control_slots',
        'nFastNetworkControlSlots': 'n_fast_network_control_slots',
        'networkIdentWidth': 'network_ident_width',
    }

    @classmethod
    def from_json(cls, json_str: str) -> 'FMVPUParams':
        """Create FMVPUParams from JSON string with camelCase field names."""
        data = json.loads(json_str)
        return cls.from_dict(data)

    @classmethod
    def from_file(cls, filename: str) -> 'FMVPUParams':
        """Create FMVPUParams from JSON file with camelCase field names."""
        with open(filename, 'r', encoding='utf-8') as f:
            data = json.load(f)
        return cls.from_dict(data)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'FMVPUParams':
        """Create FMVPUParams from dictionary with camelCase field names."""
        # Convert camelCase keys to snake_case
        converted_data = {}
        for camel_key, snake_key in cls._FIELD_MAPPING.items():
            if camel_key in data:
                converted_data[snake_key] = data[camel_key]
            else:
                raise KeyError(f"Missing required field '{camel_key}' in JSON data")
        
        return cls(**converted_data)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary with camelCase field names for JSON serialization."""
        # Convert snake_case keys to camelCase
        result = {}
        for camel_key, snake_key in self._FIELD_MAPPING.items():
            result[camel_key] = getattr(self, snake_key)
        
        return result

    def to_json(self, indent: int = None) -> str:
        """Convert to JSON string with camelCase field names."""
        return json.dumps(self.to_dict(), indent=indent)

    def to_file(self, filename: str, indent: int = 2) -> None:
        """Write to JSON file with camelCase field names."""
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(self.to_dict(), f, indent=indent)

    @property
    def words_per_channel_slow_network_control(self) -> int:
        """Calculate words per channel slow network control."""
        from control_structures import ChannelSlowControl
        bits = ChannelSlowControl.width_bits(self)
        return (bits + self.width - 1) // self.width

    @property
    def words_per_general_slow_network_control(self) -> int:
        """Calculate words per general slow network control."""
        from control_structures import GeneralSlowControl
        bits = GeneralSlowControl.width_bits(self)
        return (bits + self.width - 1) // self.width

    @property
    def words_per_channel_fast_network_control(self) -> int:
        """Calculate words per channel fast network control."""
        from control_structures import ChannelFastControl
        bits = ChannelFastControl.width_bits(self)
        return (bits + self.width - 1) // self.width

    @property
    def words_per_general_fast_network_control(self) -> int:
        """Calculate words per general fast network control."""
        from control_structures import GeneralFastControl
        bits = GeneralFastControl.width_bits(self)
        return (bits + self.width - 1) // self.width

    @property
    def words_per_slow_network_control_slot(self) -> int:
        """Calculate words per slow network control slot (rounded up to power of 2)."""
        raw_words = self.words_per_general_slow_network_control + self.n_channels * self.words_per_channel_slow_network_control
        return 1 << math.ceil(math.log2(raw_words))

    @property
    def words_per_fast_network_control_slot(self) -> int:
        """Calculate words per fast network control slot (rounded up to power of 2)."""
        raw_words = self.words_per_general_fast_network_control + self.n_channels * self.words_per_channel_fast_network_control
        return 1 << math.ceil(math.log2(raw_words))

    @property
    def fast_network_control_offset(self) -> int:
        """Calculate offset for fast network control (after DDM space and slow control slots)."""
        # Calculate control memory start address (after DDM space)
        ddm_max_addr = self.ddm_bank_depth * self.ddm_n_banks
        control_mem_start = 1 << clog2(ddm_max_addr) if ddm_max_addr > 0 else 0
        # Add slow control slots
        return control_mem_start + self.n_slow_network_control_slots * self.words_per_slow_network_control_slot


